resolver_mac.c

```diff
@@ -33,40 +33,58 @@ typedef struct proxy_resolver_mac_s {
+        length = 7;
+    } else if (CFEqual(proxy_type, kCFProxyTypeHTTPS)) {
+        // "HTTPS Proxy" on macOS means "use CONNENCT verb for a https:// URL", the proxy still should be an HTTP proxy.
+        scheme = "http://";
+        length = 7;
+    } else if (CFEqual(proxy_type, kCFProxyTypeSOCKS)) {
+        scheme = "socks://";
+        length = 8;
+    } else if (CFEqual(proxy_type, kCFProxyTypeFTP)) {
+        scheme = "ftp://";
+        length = 6;
+    }
+    if (advance_len)
+        *advance_len += length;
+    return scheme;
+}
+
 static void proxy_resolver_mac_auto_config_result_callback(void *client, CFArrayRef proxy_array, CFErrorRef error) {
     proxy_resolver_mac_s *proxy_resolver = (proxy_resolver_mac_s *)client;
     if (error) {
         // Get error code
         proxy_resolver->error = CFErrorGetCode(error);
     } else {
         // Convert proxy array into PAC file return format
-        size_t proxy_count = CFArrayGetCount(proxy_array);
-        size_t max_list = proxy_count * MAX_PROXY_URL + 1;
+        const size_t proxy_count = CFArrayGetCount(proxy_array);
+        // The actual limit is on one char less, we allow str* functions to use an additional char and then we check if
+        // a result list lenght exceeds the actual limit: if (lest_len >= max_list).
+        const size_t max_list = proxy_count * MAX_PROXY_URL + 1;
         size_t list_len = 0;
 
-        proxy_resolver->list = (char *)calloc(max_list, sizeof(char));
+        proxy_resolver->list = (char *)calloc(max_list + 1, sizeof(char));  // +1 for null terminator
 
         // Enumerate through each proxy in the array
         for (size_t i = 0; proxy_resolver->list && i < proxy_count; i++) {
             CFDictionaryRef proxy = CFArrayGetValueAtIndex(proxy_array, i);
             CFStringRef proxy_type = (CFStringRef)CFDictionaryGetValue(proxy, kCFProxyTypeKey);
 
             // Copy type of connection
-            if (CFEqual(proxy_type, kCFProxyTypeNone)) {
-                strncat(proxy_resolver->list, "direct://", max_list - list_len - 1);
-                list_len += 9;
-            } else if (CFEqual(proxy_type, kCFProxyTypeHTTP)) {
-                strncat(proxy_resolver->list, "http://", max_list - list_len - 1);
-                list_len += 7;
-            } else if (CFEqual(proxy_type, kCFProxyTypeHTTPS)) {
-                strncat(proxy_resolver->list, "https://", max_list - list_len - 1);
-                list_len += 8;
-            } else if (CFEqual(proxy_type, kCFProxyTypeSOCKS)) {
-                strncat(proxy_resolver->list, "socks://", max_list - list_len - 1);
-                list_len += 8;
-            } else if (CFEqual(proxy_type, kCFProxyTypeFTP)) {
-                strncat(proxy_resolver->list, "ftp://", max_list - list_len - 1);
-                list_len += 6;
+            const size_t len = list_len;
+            const char *scheme = proxy_resolver_mac_get_proxy_scheme(proxy_type, &list_len);
```

Comment:
Let's remove `len` variable and `list_len` pass into `get_proxy_scheme`. In `if` statement below it can just be done `max_list - list_len` and then list_len += strlen(scheme)`.

Reply:
Removed.